core/types/calls.cc

```diff
@@ -2661,19 +2661,16 @@ class Magic_callWithBlock : public IntrinsicMethod {
         }
 
         {
+            auto nonNilPassedInBlockType = Types::dropNil(gs, passedInBlockType);
+            auto passedInBlockReturnType = Types::getProcReturnType(gs, nonNilPassedInBlockType);
             auto it = &dispatched;
             while (it != nullptr) {
                 if (it->main.method.exists()) {
-                    const auto &methodArgs = it->main.method.data(gs)->arguments;
-                    ENFORCE(!methodArgs.empty());
-                    const auto &bspec = methodArgs.back();
-                    ENFORCE(bspec.flags.isBlock);
-
-                    auto bspecType = bspec.type;
-                    if (bspecType) {
+                    const auto &blockReturnType = it->main.blockReturnType;
+                    if (blockReturnType) {
                         // TODO(jez) How should this interact with highlight untyped?
                         // This subtype check is here to discover the correct generic bounds.
-                        Types::isSubTypeUnderConstraint(gs, *constr, passedInBlockType, bspecType,
```

Comment:
Sorbet currently allows the parameter types of `blk` in `foo(&blk)` to accumulate constraints onto the `T.type_parameter`'s defined on `foo`'s `T.proc.params`. But that's not how `foo do |x| blk.call(x) end` would have worked—written like that, Sorbet would only allow accumulating constraints against `T.type_parameter`'s of `foo` if those type parameters occur in the block return type (not params).

The fact that Sorbet worked like this was hiding a type error in pay-server.

in general, the idea behind this `call-with-block` code should be to approximate the logic in `BlockReturn` from processBinding, and `BlockReturn` only uses calls `isSubTypeUnderConstraint` on `blockReturnType`, not on the whole `bspecType`.